### PR TITLE
Add test for MergeEnvFileAndOSEnv

### LINKED ISSUE #1884: Add unittest for MergeEnvFileAndOSEnv
### What is the problem you're trying to solve

There is currently a GenerateEnvs function in `cmd/nerdctl/run.go`, which is used in commands, `run` and `exec`. This function may also be used elsewhere in the future. We can probably extract it to flagutil.

https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/run.go#L1314-L1334

### Describe the solution you'd like

Extract this function and its dependencies into flagutil and add some tests for it.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by suyanhanx**
close #1884 

Signed-off-by: Han Xu <suyanhanx@gmail.com>

### CODE REVIEW

File: pkg/flagutil/flagutil_test.go [hunk 1]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+			if actual != nil {
+				v := actual[0]
+				assert.Equal(t, v, tc.expected)
+			}
+		})
+	}
+}
+
+// Test MergeEnvFileAndOSEnv for a file with a line exceeding bufio.MaxScanTokenSize.
+func TestParseEnvFileLineTooLongFile(t *testing.T) {
```
Comment (djdongjin) [inline pkg/flagutil/flagutil_test.go:275]: can you update this to keep comment/func name/func being tested consistent?

File: pkg/flagutil/flagutil_test.go [hunk 2]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+}
+
+// TestMergeEnvFileAndOSEnv test merging variables from env-file and env.
+func TestMergeEnvFileAndOSEnv(t *testing.T) {
+	content := `HOME`
+	tmpFile := tmpFileWithContent(t, content)
+
+	variables, err := MergeEnvFileAndOSEnv([]string{tmpFile}, []string{"PATH"})
+	if nil != err {
+		t.Fatal("There must not be any error")
```
Comment (djdongjin) [inline pkg/flagutil/flagutil_test.go:305]: -> `t.Fatalf("unexpected error: %v", err)`

File: pkg/flagutil/flagutil_test.go [hunk 3]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+and_underscore=working too`
+	// Adding a newline + a line with pure whitespace.
+	// This is being done like this instead of the block above
+	// because it's common for editors to trim trailing whitespace
+	// from lines, which becomes annoying since that's the
+	// exact thing we need to test.
+	content += "\n    \t  "
+	tmpFile := tmpFileWithContent(t, content)
+
+	lines, err := parseEnvVars([]string{tmpFile})
```
[the remark this answers is not in the document]
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:118]: Yes. That's what I'm considering doing.

File: pkg/flagutil/flagutil_test.go [hunk 4]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+	content := `HOME`
+	tmpFile := tmpFileWithContent(t, content)
+
+	variables, err := MergeEnvFileAndOSEnv([]string{tmpFile}, []string{"PATH"})
+	if nil != err {
+		t.Fatal("There must not be any error")
+	}
+
+	if len(variables) != 2 {
+		t.Fatal("Variables from env-file and flag env should be merged.")
```
[the remark this answers is not in the document]
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:309]: Ah, this is the beginning of a sentence.

File: pkg/flagutil/flagutil_test.go [hunk 5]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+	if err == nil {
+		t.Fatal("ParseEnvFile succeeded; expected failure")
+	}
+	if _, ok := errors.Unwrap(err).(*os.PathError); !ok {
+		t.Fatalf("Expected a PathError, got [%v]", err)
+	}
+}
+
+// TestValidateEnv will test the validate function's correctness.
+func TestValidateEnv(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:163]: Sure.

File: pkg/flagutil/flagutil_test.go (hunk 1 above)
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:275]: Done.

File: pkg/flagutil/flagutil_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (djdongjin) [reply pkg/flagutil/flagutil_test.go:309]: I think usually in error messages/logs the 1st char doesn't need to be uppercased and the period at the end is also not needed.

[the remark this answers is not in the document]
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:309]: Tried to read some of the code-style documentation. Your thought makes sense. Will update them.

[the remark this answers is not in the document]
Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go]: Done.

**Review by djdongjin — APPROVED** [review]

File: pkg/flagutil/flagutil_test.go [hunk 6]
```diff
@@ -73,3 +97,223 @@ func TestReplaceOrAppendEnvValues(t *testing.T) {
...
+func TestParseEnvFileNonExistentFile(t *testing.T) {
+	_, err := parseEnvVars([]string{"foo_bar_baz"})
+	if err == nil {
+		t.Fatal("ParseEnvFile succeeded; expected failure")
+	}
+	if _, ok := errors.Unwrap(err).(*os.PathError); !ok {
+		t.Fatalf("Expected a PathError, got [%v]", err)
+	}
+}
+
```
Comment (fahedouch) [inline pkg/flagutil/flagutil_test.go:161]: is it possible to merge  `TestParseEnvFileGoodFile` , `TestParseEnvFileEmptyFile` and `TestParseEnvFileNonExistentFile` and use cases instead

Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:161]: I think the current way of doing tests is more readable.

File: pkg/flagutil/flagutil_test.go [hunk 7]
```diff
@@ -17,15 +17,39 @@
...
 	"sort"
+	"strings"
 	"testing"
 
 	"gotest.tools/v3/assert"
 )
 
-func TestReplaceOrAppendEnvValues(t *testing.T) {
+// tmpFileWithContent will create a temp file with given content.
+func tmpFileWithContent(t *testing.T, content string) string {
```
Comment (AkihiroSuda) [inline pkg/flagutil/flagutil_test.go:34]: Why not use `t.TempDir`

Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:34]: It looks like `t.TempDir` returns a path. Do you mean I should use this path for the going test?

Comment (AkihiroSuda) [reply pkg/flagutil/flagutil_test.go:34]: yes

Comment (suyanhanx) [reply pkg/flagutil/flagutil_test.go:34]: Make sense. Done.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**